src/substitutions.ts

```diff
@@ -0,0 +1,317 @@
+                if (room && room.SlackTeamId === teamId) {
+                    body = body.replace(alias, `<#${room.SlackChannelId!}>`);
+                }
+                match = regex.exec(htmlString);
+            }
+        }
+
+        // convert @room to @channel
+        body = body.replace("@room", "@channel");
+
+        // Strip out any language specifier on the code tags, as they are not supported by slack.
+        body = body.replace(/```[\w*]+\n/g, "```\n");
+
+        // Note: This slower plainTextSlackMentions call is only used if there is not a pill in the message,
+        // meaning if we can we use the much simpler pill subs rather than this.
+        const modifiedBody = await plainTextSlackMentions(main, body, event.room_id);
+
+// tslint:disable-next-line: no-any
+        const ret: any = {
+            link_names: 1,  // This no longer works for nicks but is needed to make @channel work.
+            text: modifiedBody,
+            username: event.user_id,
+        };
+        if (event.content.msgtype === "m.image" && event.content.url.indexOf("mxc://") === 0) {
+            const url = main.getUrlForMxc(event.content.url);
+            delete ret.text;
+            ret.attachments = [{
+                fallback: modifiedBody,
+                image_url: url,
+            }];
+        } else if (event.content.msgtype === "m.file" && event.content.url.indexOf("mxc://") === 0) {
+            const url = main.getUrlForMxc(event.content.url);
+            ret.text = `<${url}|${modifiedBody}>`;
+        }
+
+        return ret;
+    }
+
+    public htmlEscape(s: string) {
+        return s.replace(/&/g, "&amp;")
+            .replace(/</g, "&lt;")
+            .replace(/>/g, "&gt;");
+    }
+
+    /**
+     * Get a mapping of the nicknames of all Slack Ghosts in the room to their slack user ids.
+     *
+     * @param {String} room_id The room id to make the maps for.
+     * @return {Promise} A mapping of display name to slack user id.
+     */
+    public async getDisplayMap(main: Main, roomId: string) {
+        const displaymap: {[name: string]: string} = {};
+        const store = main.userStore;
+        const users = await main.listGhostUsers(roomId);
+        const storeUsers: {display_name: string, id: string}[][] = await Promise.all(
+            users.map((id: string) => store.select({id})),
+        );
+        storeUsers.forEach((user) => {
+            if (user && user[0]) {
+                displaymap[user[0].display_name] = user[0].id.split("_")[2].split(":")[0];
+            }
+        });
+        return displaymap;
+    }
+
+    /**
+     * Construct a mapping of the first words of all the display names to
+     * an array of objects which map full display names to their slack
+     * user ids.
+     * i.e. {"bob": [{"bob": "U123123"}, {"bob smith": "U2891283"}]}
+     *
+     * @param {Object} displaymap A mapping of display names to slack user ids.
```

Comment:
It's not worth our time to switch to typescript and not use types. Please use types.